### PR TITLE
test(conformance): Enable smoke test in ipv6 cluster

### PR DESCRIPTION
**Pull request by sayboras**
This PR is to enable  conformance test for ipv6-only cluster.

Due to the lack of IPv6 support in github action, two tests related to 
outgoing traffic (1.1.1.1 and www.google.com) are skipped now.

I take this chance to standardise env variable in IPv4 smoke
test.

Relates #12201

Signed-off-by: Tam Mach <sayboras@yahoo.com>

**_Note:_** This PR is also having new test cases added in https://github.com/cilium/cilium/pull/12599.
I will rebase and add auto generated steps later, just mark this 
PR ready for early feedback

```release-note
test(conformance): Enable ipv6 smoke test in ipv6 cluster
```

### CODE REVIEW

File: .github/workflows/smoke-test-ipv6.yaml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        run: |
+          kind load docker-image --name chart-testing cilium/cilium:latest
+          kind load docker-image --name chart-testing cilium/operator-generic:latest
+
+      - name: Install cilium chart
+        run: |
+          helm upgrade -i cilium ./install/kubernetes/cilium \
+            --namespace kube-system \
+            --set global.nodeinit.enabled=true \
+            --set global.kubeProxyReplacement=partial \
```
Comment (brb) [inline .github/workflows/smoke-test-ipv6.yaml:55]: You can use `--set global.kubeProxyReplacement=strict`. This will allow us to to remove the `nodePort`, `externalIPs`, `hostServices` `hostPort` settings.

File: examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml [hunk 2]
```diff
@@ -0,0 +1,1338 @@
...
+metadata:
+  name: echo-b-host-headless
+spec:
+  type: ClusterIP
+  clusterIP: None
+  selector:
+    name: echo-b-host
+---
+apiVersion: v1
+kind: Service
```
Comment (brb) [inline examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml]: Once https://github.com/cilium/cilium/issues/9776 has been fixed, we can set the `ipFamily`.

File: .github/workflows/smoke-test-ipv6.yaml (hunk 1 above)
Comment (sayboras) [reply .github/workflows/smoke-test-ipv6.yaml:55]: oh nice, I didn't know about this _strict_ option. Thanks

File: .github/workflows/smoke-test.yaml [hunk 3]
```diff
@@ -60,49 +62,43 @@ jobs:
       - name: Run ${{ matrix.target.name }}
         run: |
           kubectl apply -f ${{ matrix.target.template }}
-          kubectl wait -n kube-system --for=condition=Ready --all pod --timeout=5m
+          kubectl wait -n kube-system --for=condition=Ready --all pod --timeout=${{ env.TIMEOUT }}
```
Comment (tgraf) [inline .github/workflows/smoke-test.yaml]: Can you split this into a separate commit and explain why the default timeout is changed from 5m to 2m?

File: examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml [hunk 4]
```diff
@@ -0,0 +1,1338 @@
+# Automatically generated by Makefile. DO NOT EDIT
```
Comment (tgraf) [inline examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml]: This doesn't seem to be automatically generated but a copy of the generated yaml that is then modified. This means that this will get oudated quickly as it will not receive the same fixes. Can we instead automatically generate this from the IPv4 version?

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks awesome, two small requests

File: .github/kind-config-ipv6.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        apiVersion: kubeadm.k8s.io/v1beta2
+        kind: InitConfiguration
+        nodeRegistration:
+          taints: []
+  - role: worker
+networking:
+  ipFamily: ipv6
+  disableDefaultCNI: true
+  podSubnet: "fd00:10:244::/64"
+  serviceSubnet: "fd00:10:96::/112"
```
Comment (fristonio) [inline .github/kind-config-ipv6.yaml]: Needs a new line.

File: .github/workflows/smoke-test-ipv6.yaml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            --namespace kube-system \
+            --set global.nodeinit.enabled=true \
+            --set global.kubeProxyReplacement=strict \
+            --set global.pullPolicy=Never \
+            --set global.ipv6.enabled=true \
+            --set global.ipv4.enabled=false \
+            --set global.tunnel=disabled \
+            --set global.autoDirectNodeRoutes=true
+
+          kubectl wait -n kube-system --for=condition=Ready --all pod --timeout=5m
```
Comment (fristonio) [inline .github/workflows/smoke-test-ipv6.yaml]: I think you missed using `env.TIMEOUT` here.

File: .github/workflows/smoke-test-ipv6.yaml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+          # To make sure that cilium CRD is available (default timeout is 5m)
```
Comment (fristonio) [inline .github/workflows/smoke-test-ipv6.yaml:68]: Are we configuring the default timeout to be 5m or 2m?

File: .github/workflows/smoke-test-ipv6.yaml [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        if: ${{ failure() }}
+        run: |
+          for p in $(kubectl get pods --no-headers -o custom-columns=":metadata.name"); do
+              echo ---------------------------------------------------------------------------------
+              echo $p
+              echo ---------------------------------------------------------------------------------
+              kubectl describe pod $p
+              echo ---------------------------------------------------------------------------------
+              kubectl logs $p --since=${{ env.LOG_TIME }}
+          done
```
Comment (fristonio) [inline .github/workflows/smoke-test-ipv6.yaml]: Needs a new line at the end.

**Review by fristonio — CHANGES_REQUESTED** [review]
Overall looks great :rocket:

File: examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml (hunk 4 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml]: agreed, I wait for #12599 then will rebase and add script to generate this.

File: .github/workflows/smoke-test-ipv6.yaml (hunk 6 above)
Comment (sayboras) [reply .github/workflows/smoke-test-ipv6.yaml]: good catch :+1:

File: .github/workflows/smoke-test-ipv6.yaml (hunk 7 above)
Comment (sayboras) [reply .github/workflows/smoke-test-ipv6.yaml:68]: oh thanks, the comment seems outdated as there was recent change in operator CRD. Let me clean it up.

File: .github/workflows/smoke-test-ipv6.yaml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+          # To make sure that cilium CRD is available (default timeout is 5m)
+          # https://github.com/cilium/cilium/blob/master/operator/crd.go#L34
+          kubectl wait --for condition=Established crd/ciliumnetworkpolicies.cilium.io --timeout=${{ env.TIMEOUT }}
+
+      - name: Run conformance test (e.g. connectivity check without external 1.1.1.1 and www.google.com)
+        run: |
+          kubectl apply -f ${{ env.CONFORMANCE_TEMPLATE }}
+          kubectl wait --for=condition=Available --all deployment --timeout=${{ env.TIMEOUT }}
+
+      - name: Dump cilium related logs and events
```
Comment (joestringer) [inline .github/workflows/smoke-test-ipv6.yaml:76]: :thinking: How do we keep these consistent across different checks?

(Worth also noting that both https://github.com/cilium/cilium/pull/12689 and https://github.com/cilium/cilium/pull/12599 will likely extend the existing GH output we gather for debugging)

File: .github/workflows/smoke-test.yaml (hunk 3 above)
Comment (sayboras) [reply .github/workflows/smoke-test.yaml]: makes sense :+1:

File: .github/workflows/smoke-test-ipv6.yaml (hunk 9 above)
Comment (sayboras) [reply .github/workflows/smoke-test-ipv6.yaml:76]: @joestringer good point, what do think about add one more label (e.g. category) so that we can just filter based on label ?

Comment (sayboras) [reply .github/workflows/smoke-test-ipv6.yaml:76]: it could be based on the logical group in generation script.

File: examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml (hunk 2 above)
Comment (aanm) [reply examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml]: It does not hurt to set it before #9776 is fixed.

File: .github/workflows/smoke-test-ipv6.yaml (hunk 9 above)
Comment (joestringer) [reply .github/workflows/smoke-test-ipv6.yaml:76]: I'm not sure whether I follow. How would we configure / use the label?

My concern is that this is a brand new YAML for IPv6 and it may diverge from the IPv4 version. Maybe my concern is overblown if there's just two versions of the YAMLs, we can just ensure that we update both. I'd just want to make sure that any debuggability improvements we make will apply to both versions of the check.

Comment (joestringer) [reply .github/workflows/smoke-test-ipv6.yaml:76]: Hmm, actually I just hit this type of problem with #12599 where I *thought* #12689 would help by printing the info I need when it fails, however I updated the "dump logs" section for the wrong check so I get no more information than before. As always, easy fix but also easy to miss on initial submission :-)

**Review by aanm — APPROVED** [review]

File: examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml [hunk 10]
```diff
@@ -0,0 +1,1338 @@
...
+    - toEndpoints:
+        - matchLabels:
+            k8s:io.kubernetes.pod.namespace: openshift-dns
+            k8s:dns.operator.openshift.io/daemonset-dns: default
+      toPorts:
+        - ports:
+            - port: "5353"
+              protocol: UDP
+---
+#apiVersion: apps/v1
```
Comment (nebril) [inline examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml]: Can we remove this leftover?

**Review by nebril — COMMENTED** [review]
Cool stuff! Let's make the yaml generated and ship it!

File: .github/workflows/smoke-test-ipv6.yaml (hunk 1 above)
Comment (rolinh) [reply .github/workflows/smoke-test-ipv6.yaml:55]: Mmh, I don't think that `kubeProxyReplacement=strict` plays well with `kind`. I'd stick to `partial` for this one.

Comment (joestringer) [reply .github/workflows/smoke-test-ipv6.yaml:55]: @Rolinh Is it kind or is it the underlying kernel that kind is running on? I'd +1 on `strict` if it works in CI. Not sure what kernel we get there.

Comment (rolinh) [reply .github/workflows/smoke-test-ipv6.yaml:55]: @joestringer Not sure what the real issue is but if I enable `strict` on a `kind` cluster on my machine (with kernel `5.7.10`) I run into all sort of issues such as coredns pods not getting into ready-state, cilium pod in crashloop trying to reach the kube api server, etc.

File: .github/workflows/smoke-test-ipv6.yaml (hunk 9 above)
Comment (sayboras) [reply .github/workflows/smoke-test-ipv6.yaml:76]: thanks, I just rebased with your changes to make it standardized.

File: examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml (hunk 4 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml]: added cue script to generate this file :)

File: examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml (hunk 10 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml]: thanks, done :)

File: .github/workflows/smoke-test-ipv6.yaml [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        run: |
+          helm upgrade -i cilium ./install/kubernetes/cilium \
+            --namespace kube-system \
+            --set global.nodeinit.enabled=true \
+            --set global.kubeProxyReplacement=partial \
+            --set global.hostServices.enabled=false \
+            --set global.externalIPs.enabled=true \
+            --set global.nodePort.enabled=true \
+            --set global.hostPort.enabled=true \
+            --set config.bpfMasquerade=false \
```
Comment (sayboras) [inline .github/workflows/smoke-test-ipv6.yaml:60]: I added bpfMasquerade flag to make it in sync with ipv4 conformance test. Thanks to @joestringer

**Comment by sayboras** [comment]
test-me-please

File: examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml (hunk 2 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/connectivity-check-ipv6.yaml]: sure, I tried to achieve this based on flag ipFamly  from cue command, but the code is quite ugly, maybe I need to read cuelang docs a little bit more. Let me do it as follow up.

File: examples/kubernetes/connectivity-check/.gitignore [hunk 12]
```diff
@@ -0,0 +1 @@
+servers.yaml
```
Comment (fristonio) [inline examples/kubernetes/connectivity-check/.gitignore:1]: Thanks for this.

File: examples/kubernetes/connectivity-check/Makefile [hunk 13]
```diff
@@ -30,6 +32,10 @@ $(DEFAULT_OUT): $(SRC)
 	@echo '# Automatically generated by Makefile. DO NOT EDIT' > $(DEFAULT_OUT)
 	$(QUIET)$(CUE) dump > $(DEFAULT_OUT)
 
+$(IPv6_ONLY_OUT): $(SRC)
+	@echo '# Automatically generated by Makefile. DO NOT EDIT' > $(IPv6_ONLY_OUT)
+	$(QUIET)$(CUE) -t ipFamily=IPv6 dump > $(IPv6_ONLY_OUT)
```
Comment (fristonio) [inline examples/kubernetes/connectivity-check/Makefile]: Sorry, I missed out on the review of Joe's PR on Cue usage in connectivity-checks. But I think this line should append the cue dump `$(QUIET)$(CUE) -t ipFamily=IPv6 dump >> $(IPv6_ONLY_OUT)` to keep the autogenerated header on the top of the file, right?

**Review by fristonio — CHANGES_REQUESTED** [review]

File: examples/kubernetes/connectivity-check/Makefile (hunk 13 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/Makefile]: this is very good catch, thanks a lot :100: 

Let me take this chance to update for other targets as well.

File: examples/kubernetes/connectivity-check/Makefile [hunk 14]
```diff
@@ -28,24 +30,28 @@ all: $(ALL_TARGETS)
 
 $(DEFAULT_OUT): $(SRC)
 	@echo '# Automatically generated by Makefile. DO NOT EDIT' > $(DEFAULT_OUT)
-	$(QUIET)$(CUE) dump > $(DEFAULT_OUT)
+	$(QUIET)$(CUE) dump >> $@
+
+$(IPv6_ONLY_OUT): $(SRC)
+	@echo '# Automatically generated by Makefile. DO NOT EDIT' > $(IPv6_ONLY_OUT)
+	$(QUIET)$(CUE) -t ipFamily=IPv6 dump >> $@
```
Comment (sayboras) [inline examples/kubernetes/connectivity-check/Makefile:37]: for reviewer, there was a mixed usage of $@ and $(OUT_FILE) in this Makefile. As I am changing this file, I just change all to $@ for standardization.

> $@
> The file name of the target of the rule. If the target is an archive member, then ‘$@’ is the name of the archive file. In a pattern rule that has multiple targets (see Introduction to Pattern Rules), ‘$@’ is the name of whichever target caused the rule’s recipe to be run.

File: examples/kubernetes/connectivity-check/Makefile [hunk 15]
```diff
@@ -28,24 +30,28 @@ all: $(ALL_TARGETS)
...
 
 $(HOSTPORT_OUT): $(SRC)
 	@echo '# Automatically generated by Makefile. DO NOT EDIT' > $(HOSTPORT_OUT)
-	$(QUIET)$(CUE) -t component=all dump > $(HOSTPORT_OUT)
+	$(QUIET)$(CUE) -t component=all dump >> $@
```
Comment (sayboras) [inline examples/kubernetes/connectivity-check/Makefile:42]: For reviewer, there was a mixed usage of $@ and $(OUT_FILE) in this Makefile. As I am changing this file, I just change all to $@ for standardization.

$@
The file name of the target of the rule. If the target is an archive member, then ‘$@’ is the name of the archive file. In a pattern rule that has multiple targets (see Introduction to Pattern Rules), ‘$@’ is the name of whichever target caused the rule’s recipe to be run.

Comment (fristonio) [reply examples/kubernetes/connectivity-check/Makefile:42]: While we are at this, I am not sure why there is an inconsistent usage of `@` and `$(QUIET)` in the Makefile. Is there a particular reason we are using a mix of `@` and `$(QUIET)` here @joestringer ?

File: examples/kubernetes/connectivity-check/Makefile (hunk 13 above)
Comment (fristonio) [reply examples/kubernetes/connectivity-check/Makefile]: That would be great. :vulcan_salute:

File: examples/kubernetes/connectivity-check/Makefile (hunk 15 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/Makefile:42]: I think we can control the baviour based of `$(QUIET)` based on `$(V)` (e.g. export V=0), meanwhile, `@` behavior is unchangable.

**Review by nebril — APPROVED** [review]

File: examples/kubernetes/connectivity-check/Makefile (hunk 15 above)
Comment (fristonio) [reply examples/kubernetes/connectivity-check/Makefile:42]: But I still don't understand if a mix of these makes sense here. :sweat_smile:
Anyways the changes look good to me overall and this is not a problem.

**Review by fristonio — APPROVED** [review]

File: examples/kubernetes/connectivity-check/Makefile (hunk 15 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/Makefile:42]: Re: inconsistent usage, for some cases I figure that when you want to debug or understand the process occurring, you would want to see the commands being run. Typically for `echo` commands (especially when printing output to the terminal), this is not useful because you end up with duplicate printing: Once in the command output and once in the stdout.

Though that said I almost exclusively turn $(QUIET) on so I forget what the build looks like without it :-)

File: examples/kubernetes/connectivity-check/Makefile [hunk 16]
```diff
@@ -5,6 +5,8 @@ include ../../../Makefile.defs
 include ../../../Makefile.quiet
 
 DEFAULT_OUT := connectivity-check.yaml
+# Same as connectivity-check.yaml, but without 1.1.1.1 and google.com
+IPv6_ONLY_OUT := connectivity-check-ipv6.yaml
```
Comment (joestringer) [inline examples/kubernetes/connectivity-check/Makefile]: nit: I wonder if it's worth calling this `internal` or `cluster` or similar given that the actual implementation is not changing to guarantee ipv4 / ipv6, it's just turning on/off external checks.

**Review by joestringer — COMMENTED** [review]
CUE usage looks good, although I think it's a little misleading to call this "ipv6" at the connectivity check layer since all it does is exclude the external checks. If IPv4 is available, it will still use IPv4.

I think if we were to specifically call out IPv4/IPv6, we should consider passing IP family parameters to the curl commands to ensure IPv6 is used. That said I don't think this is necessary for the goals of this PR; in the context of the IPv6 GH smoke test if we're disabling IPv4 then it should only perform IPv6 checks (because any IPv4 connections will be dropped).

I didn't look closely at the smoke test YAML, I figured others have more experience with what that should look like. Feel free to re-request my review if you are not getting sufficient eyes on that area.

**Comment by sayboras** [comment]
That is really the better naming, thanks :). Let me make changes

File: examples/kubernetes/connectivity-check/Makefile (hunk 16 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/Makefile]: This is really better naming, let me make changes for this.

Comment (sayboras) [reply examples/kubernetes/connectivity-check/Makefile]: I have made changes, please help to take another :eyes: whenever you have time. Thanks :)

File: examples/kubernetes/connectivity-check/main_tool.cue [hunk 17]
```diff
@@ -75,8 +76,20 @@ ccCommand: {
...
 	}
 
+	task: filterTraffic: {
+        if #flags.traffic == "internal" {
+            resources: [ for x in task.filterName.resources if x.metadata.labels.traffic != "external"  {x}]
+        }
+		if #flags.traffic == "external" {
+            resources: [ for x in task.filterName.resources if x.metadata.labels.traffic != "internal"  {x}]
+        }
+		if #flags.traffic == "any" {
```
Comment (joestringer) [inline examples/kubernetes/connectivity-check/main_tool.cue]: nit: There looks like there's something funny going on here with the spacing.

One of my TODO's is to add in `cue fmt` to help format cue files in a consistent way, I suspect that'll modify the file here.

File: examples/kubernetes/connectivity-check/main_tool.cue [hunk 18]
```diff
@@ -75,8 +76,20 @@ ccCommand: {
 		}
 	}
 
+	task: filterTraffic: {
+        if #flags.traffic == "internal" {
+            resources: [ for x in task.filterName.resources if x.metadata.labels.traffic != "external"  {x}]
+        }
+		if #flags.traffic == "external" {
+            resources: [ for x in task.filterName.resources if x.metadata.labels.traffic != "internal"  {x}]
+        }
```
Comment (joestringer) [inline examples/kubernetes/connectivity-check/main_tool.cue]: you could probably simplify this to one if condition:
```suggestion
        if #flags.traffic != "any" {
            resources: [ for x in task.filterName.resources if x.metadata.labels.traffic != #flags.traffic  {x}]
        }
```

File: examples/kubernetes/connectivity-check/network.cue [hunk 19]
```diff
@@ -7,8 +7,13 @@ package connectivity_check
 _networkCheck: {
 	metadata: labels: component: "network-check"
 }
+
+_trafficCheck: {
+    metadata: labels: traffic: "external"
```
Comment (joestringer) [inline examples/kubernetes/connectivity-check/network.cue]: nit: it's possible to expand the _networkCheck: section to include this, for example:

```
_networkCheck: {
    metadata: labels: {
        component: "network-check"
        traffic: "external"
    }
}
```

**Review by joestringer — CHANGES_REQUESTED** [review]
Just a few style nit comments above, but otherwise LGTM.

File: examples/kubernetes/connectivity-check/main_tool.cue (hunk 17 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/main_tool.cue]: oh nice, I think it's due to my Goland setup with tab and space. Will run cue fmt, thanks

File: examples/kubernetes/connectivity-check/network.cue (hunk 19 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/network.cue]: Not sure if we can merge this together, as `pod-to-a` will have traffic as external as well.

File: examples/kubernetes/connectivity-check/main_tool.cue (hunk 18 above)
Comment (sayboras) [reply examples/kubernetes/connectivity-check/main_tool.cue]: should it be something like below as x.metadata.labels.traffic could be "any"? We also need another default case to avoid ` undefined field "resources"` issue.

```
		if #flags.traffic != "any" {
			resources: [ for x in task.filterName.resources if x.metadata.labels.traffic == #flags.traffic || x.metadata.labels.traffic == "any" {x}]
		}
		if #flags.traffic == "any" {
			resources: task.filterName.resources
		}
```

File: examples/kubernetes/connectivity-check/main_tool.cue [hunk 20]
```diff
@@ -75,8 +76,17 @@ ccCommand: {
 		}
 	}
 
+	task: filterTraffic: {
+		if #flags.traffic != "any" {
+			resources: [ for x in task.filterName.resources if x.metadata.labels.traffic == #flags.traffic || x.metadata.labels.traffic == "any" {x}]
```
Comment (joestringer) [inline examples/kubernetes/connectivity-check/main_tool.cue:81]: I'm not sure I understand why the "any" comparison is necessary, the line towards the top of the file marks `*"any"` as default for `#flags.traffic` via the `*`, so it should always have a value..?

**Review by joestringer — APPROVED** [review]

File: examples/kubernetes/connectivity-check/main_tool.cue (hunk 20 above)
Comment (joestringer) [reply examples/kubernetes/connectivity-check/main_tool.cue:81]: ^ Not a big deal though, I'm happy with the rest of this PR as-is.

**Review by rolinh — APPROVED** [review]
